crates/sui-tool/src/lib.rs

```diff
@@ -964,6 +985,107 @@ pub async fn download_formal_snapshot(
+    // if restoring from epoch_891, we want to backfill checkpoints from end of epoch_890 to end of epoch_891
+    // (when restoring from epoch_891, node will immediately start in epoch 892, so prev_epoch == epoch_891)
+    let epoch_last_cp_seq = end_of_epoch_checkpoint_seq_nums
+        .get(epoch as usize)
+        .ok_or_else(|| anyhow!("No checkpoint sequence found for epoch {}", epoch))?;
+
+    let epoch_start_cp = if epoch == 0 {
+        0
+    } else {
+        end_of_epoch_checkpoint_seq_nums
+            .get(epoch as usize - 1)
+            .map(|cp| cp + 1)
+            .unwrap_or(0)
+    };
+    m.println(format!(
+        "Beginning transaction digest backfill for epoch: {:?}, backfilling from: {:?}..{:?}",
+        epoch, epoch_start_cp, epoch_last_cp_seq
+    ))?;
+
+    let checkpoints_to_fetch: Vec<_> = (epoch_start_cp..=*epoch_last_cp_seq).collect();
+    let num_checkpoints = checkpoints_to_fetch.len();
+
+    let progress_bar = m.add(
+        ProgressBar::new(num_checkpoints as u64).with_style(
+            ProgressStyle::with_template(
+                "[{elapsed_precise}] {wide_bar} {pos}/{len} transactions backfilled ({msg})",
+            )
+            .unwrap(),
+        ),
+    );
+
+    let client = create_remote_store_client(ingestion_url, vec![], 60)?;
+    let checkpoint_counter = Arc::new(AtomicU64::new(0));
+    let tx_counter = Arc::new(AtomicU64::new(0));
+    let cloned_checkpoint_counter = checkpoint_counter.clone();
+    let cloned_progress_bar = progress_bar.clone();
+    let start_instant = Instant::now();
+
+    tokio::spawn(async move {
+        loop {
+            if cloned_progress_bar.is_finished() {
+                break;
+            }
+            let num_checkpoints_processed = cloned_checkpoint_counter.load(Ordering::Relaxed);
+            let elapsed = start_instant.elapsed().as_secs_f64();
+            let chkpts_per_sec = if elapsed > 0.0 {
+                num_checkpoints_processed as f64 / elapsed
+            } else {
+                0.0
+            };
+            cloned_progress_bar.set_position(num_checkpoints_processed);
+            cloned_progress_bar.set_message(format!("{:.1} chkpts/sec", chkpts_per_sec));
+            tokio::time::sleep(Duration::from_millis(100)).await;
+        }
+    });
+
+    futures::stream::iter(checkpoints_to_fetch)
+        .map(|sq| CheckpointReader::fetch_from_object_store(&client, sq))
+        .buffer_unordered(concurrency)
+        .try_for_each(|checkpoint| {
+            let perpetual_db = perpetual_db.clone();
+            let tx_counter = tx_counter.clone();
+            let checkpoint_counter = checkpoint_counter.clone();
+            let checkpoint_data = checkpoint.0;
+
+            async move {
+                for tx_data in &checkpoint_data.transactions {
+                    let tx_digest = tx_data.transaction.digest();
+                    perpetual_db.insert_executed_transaction_digest(epoch, tx_digest)?;
```

Comment:
related to my prior comment, will probably be much faster to do batched writes on a per-checkpoint basis